### ISSUE TITLE
Tolls vs Vehicle Type

### ISSUE THREAD

**Issue by MotoUKRider** [abrensch/brouter#639]
Following from [this discussion on the Cruiser app group here](https://github.com/devemux86/cruiser/discussions/134#discussioncomment-7380651), is there an allowance for routing across a toll road when the OSM suggest that the vehicle in question (a motorcycle) does not pay a toll?

It seems that "Avoid Toll" is absolute, regardless of vehicle type and toll property.

My thinking is that if I have "Avoid Toll" on a motorcycle profile within Cruiser app, I should be able to route over any toll road where a motorcycle passes for free.

**Comment by devemux86** [comment]
To clarify, BRouter does not seem to include the `toll:motorcycle` in the routing data?
So this tag cannot be used in profiles, we get the message "unknown lookup name: toll:motorcycle".

For example this bridge with `toll=yes` and `toll:motorcycle=no`: https://www.openstreetmap.org/way/4063883

**Comment by afischerdev** [comment]
I hvae added this to #458 fro the next generation lookups

**Comment by devemux86** [comment]
@MotoUKRider 

This is related to the download data of BRouter (not the code).
There is a large data update that has not yet been merged here:
- https://github.com/abrensch/brouter/pull/458

**Comment by MotoUKRider** [comment]
Should this be working now?  I can't get it to work on the BRouter-test website (or within Cruiser App using "Motorcycle" profile).

<img width="1140" height="475" alt="Image" src="https://github.com/user-attachments/assets/b72d2c55-54c2-42ac-b5c5-567132efcfec" />

My [sample route](https://brouter.de/brouter-test/#map=11/51.4628/0.2462/standard&lonlats=0.280977,51.495866;0.238078,51.438169) is based on [this toll bridge crossing](https://www.openstreetmap.org/way/20003982#map=14/51.46577/0.25625) which is correctly tagged as free for motorcycles. The adjacent (northbound) tunnels are also correctly tagged.

I used attached driving profile with the following settings (which I think set the vehicle type):
```
# Car profile
assign validForCars        = false
# Motorcycle profile
assign validForMotorcycles = true
```
[driving2.brf.txt](https://github.com/user-attachments/files/28678916/driving2.brf.txt)

Although there are only 2 toll crossings that I know of here in the UK that are free for motorcycles, both would cause a person _not_ familiar with their operation a considerable detour in this situation!  More so for the rider just arrived from France who avoided tolls and wants to enjoy our eastern counties without having to drive into London and forced into taking the ferry across the Thames or using a tunnel in the city.

**Comment by devemux86** [comment]
@MotoUKRider You probably need to adjust the profile and include the new toll tag in the rules.

**Comment by MotoUKRider** [comment]
For anyone following this, I've had some success but obviously do not understand the syntax correctly (proof that users shouldn't be expected to code) as there are 2 issues presently.

Profile attached and I have commented-out the original line 64 and added 127. This seems to work perfectly for _motorcycles_ on the BRouter online planner but crashes Cruiser App when trying to use the "Routing" menu. I have to delete the file from my phone to get it running with the default internal profile. So, while it _seems_ to work, it's probably not in the right place.

The other issue is for _car_ users. Assuming an App can use this profile and switch the car/motorcycle values between `true/false` then it needs to handle both - If set as a _car_ , the toll avoid (understandably) doesn't work as I only used "motorcycle" in my line. I am guessing that I need _two_ new bits of code, placed in the vehicle access sections, but I mostly get get errors in the BRouter console (expected this, missing that etc).

I have studied the profile documentation and looked at a couple of example profiles but this isn't really my "thing". I'm a Stupid Mototcylist who is happy to get oil on his hands but not dismantle the bike. I'm sure the solution will look easy if/when it's presented!

[moto-toll.brf.txt](https://github.com/user-attachments/files/29241598/moto-toll.brf.txt)

**Comment by afischerdev** [comment]
@MotoUKRider 
Sorry for the delay. But here are my 5 Cents:
I splitted the rules first
```
assign is_avoided_toll_road_car  and avoid_toll toll=yes
assign is_avoided_toll_road_motorcycle  and avoid_toll not toll:motorcycle=no
```
And bring them together later on
```
assign is_avoided_toll_road
  if and validForMotorcycles is_avoided_toll_road_motorcycle then 1
  else 
  if and not validForMotorcycles is_avoided_toll_road_car then 1
  else 0
```

**Comment by MotoUKRider** [comment]
@afischerdev Thanks for the hints, that sort of helps and I'm beginning to understand it better now.

However, adding those lines into the profile seems to make a route _only on the toll section of the road_ - leaving a big gap between the Start+End markers and the start+end of the purple route line.  "Start Marker is too far from a route" is an error I also see if recalculating a previously working route with altered toll settings.

I am thinking that the 2nd line of your code is causing a problem here. It seems to be setting `is_avoided_toll_road_motorcycle` to `true` when we "avoid tolls" _and_ a road is a normal one, ie: does NOT contain `toll:motorcycle=no` - which is 99% of them!  I tried `assign is_avoided_toll_road_motorcycle and avoid_toll and toll=yes not toll:motorcycle=no` but that simply did a normal (no toll) re-route.

I'm also aware that I could be chasing my own tail by editing _both_ sections of code.

**Comment by devemux86** [comment]
A fix seems to be needed:
```
assign is_avoided_toll_road_motorcycle  and avoid_toll and toll=yes not toll:motorcycle=no
```

**Comment by afischerdev** [comment]
@MotoUKRider 
You are right, I had that several times, but I didn't check for 'normal' toll situations. Here my [current test](https://brouter.de/brouter-web/#map=14/51.4993/0.0216/standard&lonlats=0.006521,51.499448;0.011102,51.509156).

@devemux86 
Yes, that's my solution as well.

**Comment by MotoUKRider** [comment]
@afischerdev @devemux86 Thanks for this, I have it working now.  The issue was not testing my last edit near London at the Dartford Crossing, but only at the Humber Bridge where [this tiny piece of road](https://www.openstreetmap.org/way/5729061) causes a hiccup with our new code. Hence it was _always_ going to fail. (for future reference: this section did not have a `toll:motorcycle=no` tag, just a `toll=yes`)

So I now have some OSM edits to make: Checking that every little section of road with a `toll=yes` also has a `toll:motorcycle=no` where neccesary otherwise it ruins everything.  It also seems to working in Cruiser under the "Motorcycle" profile :)
